enhancements/installer/aws-lb-subnet-selection.md

```diff
@@ -0,0 +1,974 @@
+)
+```
+
+The `SubnetsConfig` field will be introduced under the TechPreview feature gate, and that the `Subnets` field
+will be initially deprecated in TechPreview. Both the new `SubnetsConfig` field and the deprecation of `Subnets` field
+will graduate to GA together. Upon graduation, the Go docs should be updated to remove the
+`// subnetsConfig is only available in TechPreview.` and change `// Deprecated in Tech Preview: Use subnetsConfig`
+to `// Deprecated: Use subnetsConfig`
+
+
+##### Installer Validation Considerations
+
+CEL validation enforces that at least one `SubnetConfig` has a `Cluster` role. This requirement exists
+because if no `Cluster` role subnets are specified, the installer would have to create the `Cluster` role subnets
+within the VPC associated with the `IngressController` role subnets. While this scenario could be valid, it is likely
+to cause confusion. Cluster admins might not realize that even though they provided a valid `SubnetsConfig`
+field, the installer will still create the `Cluster` role subnets during installation. This validation could
+be relaxed in the future, provided there is a feature request to automatically create `Cluster` role subnets, but
+manually specify `IngressController` role subnets.
+
+The existing validation for the deprecated `Subnets` field should apply to all subnets (`IngressController`
+and `Cluster`) specified in `SubnetsConfig`, such as the [existing validation](https://github.com/openshift/installer/blob/0d77aa8df5ddc68e926aa11da24a87981021b256/pkg/asset/installconfig/aws/subnet.go#L91)
+that confirms all subnets are from the same VPC.
+
+Since the AWS Cloud Controller Manager (CCM) will reject certain subnets, the installer will need to add
+additional validation for `IngressController` role subnets:
+
+- The installer should reject multiple `IngressController` role subnets in the same AZ as this will be rejected
+  by the AWS CCM.
+
+#### IngressConfig API Updates
+
+The `spec.loadBalancer.platform.aws.classicLoadBalancer.subnets` and `spec.loadBalancer.platform.aws.networkLoadBalancer.subnets`
+fields will be added to the IngressConfig to encode the default subnets for all IngressControllers. Regardless of the
+load balancer type specified in the `platform.aws.lbType` install-config field, the installer will populate both
+`networkLoadBalancer` and `classicLoadBalancer` subnet fields with the values provided in the new
+`platform.aws.subnetsConfig` install-config field. The `AWSSubnets` struct is duplicated from the IngressController CRD
+to maintain API consistency. Although the new `subnetsConfig` install-config field only supports specifying subnets by
+ID, the IngressConfig CRD will still support specifying subnets by name for completeness.
+
+```go
+// AWSIngressSpec holds the desired state of the Ingress for Amazon Web Services infrastructure provider.
+// This only includes fields that can be modified in the cluster.
+// +union
+type AWSIngressSpec struct {
```

Comment:
> what effect would you expect the awsLBType field to have vs what's specified here?

In my opinion, for Ingress Config, I think LB Type should have no direct effect/relationship/impact on the `classicLoadBalancer` and `networkLoadBalancer` fields. These fields should be independent by the fact that these CRD fields are representing a global IngressController defaulting mechanism.

I would like to be able to say "I want my IngressControllers to use a NLB as default and to also use subnet-xyz as default, regardless of the LB Type". While it could be enforced as a union, it would limit our ability to set defaults for different LB Types. Union == restricting defaults to only the default LB Type selected at install time.

Is removing the `+union` and `+unionDiscriminator` in `AWSIngressSpec` a compatible thing to do? I don't quite understand what these tags do, other than being informational, as they don't currently enforce the union (hence why we have to use CEL).

Reply:
> Is removing the +union and +unionDiscriminator in AWSIngressSpec a compatible thing to do? I don't quite understand what these tags do, other than being informational, as they don't currently enforce the union (hence why we have to use CEL

I think this will be compatible, we need to review the doc more before we make any commitments I think

Reply:
Agreed. It's worth a mention that @miheer's PR is adding this CEL for [networkLoadBalancerParameters](https://github.com/openshift/api/pull/2043/files#diff-0beb33df12dd4ecac8d28c3fe688acc3a178de15c301325e33f586207b2c748cR134).

I think that's fine for now, since I can remove it (i.e. loosing validation) later if we decide that subnets should be defaulted for all LB types of IngressControllers.